File: C/Non_Blocking/fail_memory_function.c
#include <stdio.h>
#include <mpi.h>

#define TAG 100
#define NELEMENTS 1000000

void send_vals(int dest, MPI_Request *request){
  int i;
  int send_val[NELEMENTS]; /*This is array will go out of scope when you leave 
      this function so cannot be validly be used once this function has ended
      There may or may not actually be any problem in practice but DON'T DO IT*/

  for (i=0;i<NELEMENTS;++i){send_val[i]=10;}
  MPI_Isend(send_val, NELEMENTS, MPI_INT, dest, TAG, MPI_COMM_WORLD,
      request);
}

int main(int argc, char ** argv)
{

  int rank, nproc, left, right;
  int recv_val[NELEMENTS];
  MPI_Request requests[2];
  MPI_Status statuses[2];


  MPI_Init(&argc, &argv);
  MPI_Comm_size(MPI_COMM_WORLD, &nproc);
  MPI_Comm_rank(MPI_COMM_WORLD, &rank);

  //Set up periodic domain
  left = rank - 1;
  if (left < 0) left = nproc - 1;
  right = rank + 1;
  if (right > nproc - 1) right = 0;

  send_vals(right, &requests[0]);
  MPI_Irecv(recv_val, NELEMENTS, MPI_INT, left, TAG, MPI_COMM_WORLD,
      &requests[1]);
  MPI_Waitall(2, requests, statuses);

  printf("Rank %3d got message from rank %3d of %3d\n", rank, left, 
      recv_val[0]);

  MPI_Finalize();

}

File: Fortran/Non_Blocking/fail_memory_function.f90
MODULE fns

  USE mpi
  IMPLICIT NONE

  INTEGER, PARAMETER :: tag = 100

  CONTAINS

  SUBROUTINE send_vals(dest, nels, request)
    INTEGER, INTENT(IN) :: dest, nels
    INTEGER, DIMENSION(nels) :: vals
    INTEGER, INTENT(INOUT) :: request
    INTEGER :: ierr

    vals = 10
    CALL MPI_Isend(vals, nels, MPI_INTEGER, dest, tag, &
        MPI_COMM_WORLD, request, ierr)

  END SUBROUTINE send_vals

END MODULE fns


PROGRAM non_block

  USE mpi
  USE fns
  IMPLICIT NONE

  INTEGER, PARAMETER :: n_elements = 100000

  INTEGER :: rank
  INTEGER, DIMENSION(n_elements) :: recv_val
  INTEGER :: nproc
  INTEGER :: left, right
  INTEGER, DIMENSION(2) :: requests
  INTEGER, DIMENSION(MPI_STATUS_SIZE, 2) :: statuses
  INTEGER :: ierr
  REAL :: time1, time2, dtime

  CALL MPI_Init(ierr)

  CALL MPI_Comm_size(MPI_COMM_WORLD, nproc, ierr)
  CALL MPI_Comm_rank(MPI_COMM_WORLD, rank, ierr)

  !Set up periodic domain
  left = rank - 1
  IF (left < 0) left = nproc - 1
  right = rank + 1
  IF (right > nproc - 1) right = 0

  CALL send_vals(right, n_elements, requests(1))
  CALL MPI_Irecv(recv_val, n_elements, MPI_INTEGER, left, tag, MPI_COMM_WORLD, &
        requests(2), ierr)

  CALL MPI_Waitall(2, requests, statuses, ierr)

  PRINT *,"Rank ", rank, " got minimal value from rank ", left, " of ", &
      MINVAL(recv_val)
  CALL MPI_Barrier(MPI_COMM_WORLD, ierr)
  IF (rank == 0) PRINT *, "It took ", dtime, " seconds to complete"

  CALL MPI_Finalize(ierr)

END PROGRAM non_block
